litex/soc/cores/spiopi.py

```diff
@@ -0,0 +1,1080 @@
+                      NextValue(spi_so, spi_do),
+                      NextValue(spi_dummy, has_dummy),
+                      ).Else(
+                       NextValue(spi_clk_en, 0),
+                       NextValue(spi_ack_pipe, 0),
+                       NextValue(spicount, 0),
+                       NextValue(spi_dummy, 0),
+                   )
+                   )
+        spiphy.act("REQ",
+                   If(spicount > 0,
+                      NextValue(spicount, spicount - 1),
+                      NextValue(spi_clk_en, 1),
+                      NextValue(spi_so, Cat(0, spi_so[:-1])),
+                      NextValue(spi_ack_pipe, 0),
+                      ).Elif((spicount == 0) & spi_req & ~spi_dummy,  # back-to-back transaction
+                             NextValue(spi_clk_en, 1),
+                             NextValue(spicount, 7),
+                             NextValue(spi_clk_en, 1),
+                             NextValue(spi_so, spi_do),  # reload the so register
+                             spi_di_load.eq(1),  # "naked" .eq() create single-cycle pulses that default back to 0
+                             NextValue(spi_ack_pipe, 1),
+                             NextValue(spi_dummy, has_dummy),
+                             ).Elif((spicount == 0) & ~spi_req & ~spi_dummy,  # go back to idle
+                                    spi_di_load.eq(1),
+                                    NextValue(spi_ack_pipe, 1),
+                                    NextValue(spi_clk_en, 0),
+                                    NextState("RESET"),
+                                    ).Elif((spicount == 0) & spi_dummy,
+                                           spi_di_load.eq(1),
+                                           NextValue(spicount, self.config.fields.dummy),
+                                           NextValue(spi_clk_en, 1),
+                                           NextValue(spi_ack_pipe, 0),
+                                           NextValue(spi_so, 0),  # do a dummy with '0' as the output
+                                           NextState("DUMMY"),
+                                           )  # this actually should be a fully defined situation, no "Else" applicable
+                   )
+        spiphy.act("DUMMY",
+                   If(spicount > 1,  # instead of doing dummy-1, we stop at count == 1
+                      NextValue(spicount, spicount - 1),
+                      NextValue(spi_clk_en, 1),
+                      ).Elif(spicount <= 1 & spi_req,
+                             NextValue(spi_clk_en, 1),
+                             NextValue(spicount, 7),
+                             NextValue(spi_so, spi_do),  # reload the so register
+                             NextValue(spi_ack_pipe, 1),  # finally ack the cycle
+                             NextValue(spi_dummy, has_dummy),
+                             ).Else(
+                       NextValue(spi_clk_en, 0),
```

Comment:
Here is an example of strange indentation

Reply:
oh geez...i see it now. I did a copy/paste to a new file before merging to not break the legacy spi stuff. Pycharm went and auto-indented a bunch of areas. I'll fix that now.